kernel/src/fs/cgroupfs/systree_node.rs

```diff
@@ -487,18 +641,43 @@ inherit_sys_branch_node!(CgroupNode, fields, {
                     .ok_or(Error::InvalidOperation)?;
 
                 with_process_cgroup_locked(pid, |target_process, cgroup_membership| {
-                    // TODO: According to the "no internal processes" rule of cgroupv2
-                    // (Ref: https://man7.org/linux/man-pages/man7/cgroups.7.html),
-                    // if the cgroup node has enabled some controllers like "memory", "io",
-                    // it is forbidden to bind a process to an internal cgroup node.
                     cgroup_membership.move_process_to_node(target_process, self)
                 })?;
 
                 Ok(len)
             }
+            "cgroup.subtree_control" => {
+                let (activate_set, deactivate_set, len) = read_subtree_control_from_reader(reader)?;
+
+                let parent_node = self.cgroup_parent().ok_or(Error::IsDead)?;
+                let parent_controller = parent_node.controller().lock();
+                let mut current_controller = self.controller.lock();
+
+                self.with_inner(|processes| {
+                    // According to "no internal processes" rule of cgroupv2, if a non-root
+                    // cgroup node has bound processes, it cannot activate any sub-control.
+                    //
+                    // Reference: <https://man7.org/linux/man-pages/man7/cgroups.7.html>
+                    if !processes.is_empty() {
+                        return Err(Error::ResourceUnavailable);
+                    }
+
+                    for ctrl_type in activate_set.iter_types() {
+                        current_controller.activate(ctrl_type, self, Some(&parent_controller))?;
+                    }
+                    for ctrl_type in deactivate_set.iter_types() {
+                        current_controller.deactivate(ctrl_type, self)?;
+                    }
+
+                    Ok(len)
+                })
+                .ok_or(Error::IsDead)?
+            }
+            // TODO: Add support for writing other attributes.
             _ => {
-                // TODO: Add support for writing other attributes.
-                Err(Error::AttributeError)
+                let controller = self.controller.lock();
+                self.with_inner(|_| controller.write_attr(name, reader))
```

Comment:
Why do we need to hold the controller lock when writing to the controller attributes, yet not when reading them? Is it designed to prevent race conditions? Could you please be more specific and add some comments here?

(Actually the sub-controller can still, for example, be deactivated concurrently since we don't hold the parent's controller lock?)